### PR TITLE
Remove race conditions on inet_db access

### LINKED ISSUE #4631: inet_db terminated in 24.0-rc1
**Describe the bug**
We have been running the release candidate in production for a few days now and everything was looking great until we had a node crash, with the following log (Elixir):
```
14:01:55.259 [error] GenServer :inet_db terminating
** (ArgumentError) argument error
    (stdlib 3.15) :ets.next(:inet_cache, {'somedomain.com', :in, :cname})
    (kernel 8.0) inet_db.erl:1733: :inet_db.do_refresh_cache/4
    (kernel 8.0) inet_db.erl:1198: :inet_db.handle_info/2
    (stdlib 3.15) gen_server.erl:699: :gen_server.try_dispatch/4
    (stdlib 3.15) gen_server.erl:775: :gen_server.handle_msg/6
    (stdlib 3.15) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
Last message: :refresh_timeout
{"Kernel pid terminated",application_controller,"{application_terminated,kernel,shutdown}"}
Kernel pid terminated (application_controller) ({application_terminated,kernel,shutdown})

Crash dump is being written to: /app/tmp/erl_crash.dump...done
``` 

**To Reproduce**
With a system configured to use `inet_res` perform many concurrent DNS resolutions for an extended period of time (long enough for the cache to expire). The node on which we experienced this had been running for ~23 hours, and had cache_refresh and timeout lower than the defaults:
```
{cache_refresh, 300000}.
{timeout, 300}.
```

**Expected behavior**
No crash 🙂 

**Affected versions**
24.0-rc1

**Additional context**
CC: @josevalim as I believe this is related to some of the recent `inet_res` changes he worked on recently.

I am happy to provide any additional information that I can, and test out future patches.

### PR DESCRIPTION
**Pull request by josevalim**
Prior to this patch, a client could delete
data which would cause the server to crash.
We now change it so we ask the server to delete
the data in case of TTL=0 and avoid deleting it
altogether in other cases (we wait until the
server eventually purges it when necessary).

We also change the table structure to store
the last access separate from the dns_rr record,
allowing us to update only the latest time without
overriding any potential updated dns_rr entry.

Closes #4631.

/cc @RaimoNiskanen

### CODE REVIEW

**Comment by josevalim** [comment]
@RaimoNiskanen with this PR we are no longer updating the `.cnt` field in #dns_rr because we need to update only that field concurrently. I will send another PR to remove the `.cnt` field altogether but I didn't want to include that in this PR since this should go into the otp24 branch.

**Comment by sneako** [comment]
We have been running this in production for >22 hours now and so far everything is looking good 👍  It did take some time for the bug in #4631 to pop up in the first place, so I will report back again after another day passes.

**Comment by sneako** [comment]
Things are still looking good for us here 👍

**Comment by RaimoNiskanen** [comment]
Hello, sorry I am late to the party; I have been occupied with `socket:sendfile`...

@josevalim: have you considered, as an alternative to this patch; to use `ets:safe_fixtable(CacheDb, true|false)` over the iteration in `inet_db:do_refresh_cache/1,4`?  Would that also solve the problem?  Any performance issues?

**Comment by josevalim** [comment]
@RaimoNiskanen I see! That will be nice because we will be able to do removals for TTL=0 too.

Can I propose to try that in another PR? The biggest benefit of this PR is to move the Time outside of #dns_rr and we have safe concurrent updates. I can add safe concurrent deletions next.

If you prefer for me to tackle it here, I can do that by the end of this week!

**Comment by RaimoNiskanen** [comment]
If both changes are good, then maybe do both in this PR so to not press @sneako into too many long production tests... :-)

**Comment by RaimoNiskanen** [comment]
@josevalim, but I have no strong opinion, your call

**Comment by sneako** [comment]
> If both changes are good, then maybe do both in this PR so to not press @sneako into too many long production tests... :-)

I appreciate the sentiment but I am happy help out either way!

**Comment by josevalim** [comment]
Ok, I was able to push the safe_fixtable changes, as they were very straight-forward. @sneako, can you please give this version a try? It is the same as before + safe_fixtable. Thank you!

File: lib/kernel/src/inet_db.erl [hunk 1]
```diff
@@ -979,10 +977,10 @@ handle_call(Request, From, #state{db=Db}=State) ->
 	    do_add_rr(RR, Db, State),
 	    {reply, ok, State};
 
-	{del_rr, RR} when is_record(RR, dns_rr) ->
+	{del_rr, Key} when is_tuple(Key) ->
```
[the remark this answers is not in the document]
Comment (josevalim) [reply lib/kernel/src/inet_db.erl]: Done.

**Comment by sneako** [comment]
We have been running the latest patch for almost 6 hours now and so far, so good! I'll report back again tomorrow after it has been running for a bit longer.

**Comment by sneako** [comment]
Things have continued to be stable so far, however I did come across an unintended change in behaviour related to the recent inet_res/inet_db changes.

We use a custom :inet_tcp module [by passing the module name to gen_tcp:connect](https://erlang.org/doc/man/gen_tcp.html#type-connect_option) which randomizes the order of the list of IP addresses returned by `inet_tcp:getaddrs/1,2`. We noticed that we were no longer using all of the IP address that are being resolved.

First, I tried an older OTP version and confirmed that we saw the expected behaviour there, so I continued to investigate the issue in my patched OTP version which is based on 24.0-rc1 plus this PR.

I noticed that the first call to `inet_res:getbyname/2` did return all of the A records, but all subsequent calls (until the cache expired) only returned one record.
```
iex(3)> :inet_res.getbyname('ib.adnxs.com.', :a)
{:ok,
 {:hostent, 'ib.anycast.adnxs.com', ['g.geogslb.com', 'ib.adnxs.com'], :inet, 4,
  [
    {185, 33, 220, 240},
    {185, 33, 221, 14},
    {185, 33, 221, 15},
    {185, 33, 220, 241},
    {185, 33, 221, 89},
    {185, 33, 221, 52},
    {185, 33, 221, 91},
    {185, 33, 220, 244}
  ]}}
iex(4)> :inet_res.getbyname('ib.adnxs.com.', :a)
{:ok,
 {:hostent, 'ib.anycast.adnxs.com', ['g.geogslb.com', 'ib.adnxs.com'], :inet, 4,
  [{185, 33, 220, 244}]}}
```

So I did some tracing and noticed that for each resolved DNS record, we are inserting it into the ets inet_cache:
```
8:48:15.768052 <0.51.0> ets:insert(inet_cache, {{"ib.adnxs.com",in,cname},
 {dns_rr,"ib.adnxs.com",cname,in,-576460747,43130,"g.geogslb.com",-576460747,
         [],false}})
8:48:15.768233 <0.51.0> ets:insert(inet_cache, {{"g.geogslb.com",in,cname},
 {dns_rr,"g.geogslb.com",cname,in,-576460747,10730,"ib.anycast.adnxs.com",
         -576460747,[],false}})
8:48:15.768325 <0.51.0> ets:insert(inet_cache, {{"ib.anycast.adnxs.com",in,a},
 {dns_rr,"ib.anycast.adnxs.com",a,in,-576460747,230,
         {37,252,172,36},
         -576460747,[],false}})
8:48:15.768487 <0.51.0> ets:insert(inet_cache, {{"ib.anycast.adnxs.com",in,a},
 {dns_rr,"ib.anycast.adnxs.com",a,in,-576460747,230,
         {37,252,172,38},
         -576460747,[],false}})
8:48:15.768625 <0.51.0> ets:insert(inet_cache, {{"ib.anycast.adnxs.com",in,a},
 {dns_rr,"ib.anycast.adnxs.com",a,in,-576460747,230,
         {37,252,172,249},
         -576460747,[],false}})
8:48:15.768727 <0.51.0> ets:insert(inet_cache, {{"ib.anycast.adnxs.com",in,a},
 {dns_rr,"ib.anycast.adnxs.com",a,in,-576460747,230,
         {37,252,172,37},
         -576460747,[],false}})
8:48:15.768853 <0.51.0> ets:insert(inet_cache, {{"ib.anycast.adnxs.com",in,a},
 {dns_rr,"ib.anycast.adnxs.com",a,in,-576460747,230,
         {37,252,172,250},
         -576460747,[],false}})
8:48:15.768970 <0.51.0> ets:insert(inet_cache, {{"ib.anycast.adnxs.com",in,a},
 {dns_rr,"ib.anycast.adnxs.com",a,in,-576460747,230,
         {37,252,173,62},
         -576460747,[],false}})
8:48:15.769080 <0.51.0> ets:insert(inet_cache, {{"ib.anycast.adnxs.com",in,a},
 {dns_rr,"ib.anycast.adnxs.com",a,in,-576460747,230,
         {37,252,173,27},
         -576460747,[],false}})
8:48:15.769153 <0.51.0> ets:insert(inet_cache, {{"ib.anycast.adnxs.com",in,a},
 {dns_rr,"ib.anycast.adnxs.com",a,in,-576460747,230,
         {37,252,173,38},
         -576460747,[],false}})
```

Since we recently converted this ETS table from a bag to a set, the A records actually overwrite each other now. Therefore, I think we need to either switch back to a bag, or combine all of the A records in to one dns_rr record before writing it.

Either way, we should probably revert https://github.com/erlang/otp/commit/76a2300189193c73e2e09f6aa39218ea5eac762c and https://github.com/erlang/otp/commit/b972eb723cb850983f27695930ac5dbc6093a2c7 for now while we decide the next course of action.

**Comment by RaimoNiskanen** [comment]
Bummer!  We need to fix this.

**Comment by josevalim** [comment]
Yeah, sorry about that.

My suggestion for now is to revert 76a2300189193c73e2e09f6aa39218ea5eac762c and b972eb723cb850983f27695930ac5dbc6093a2c7 and close this PR.

**Comment by RaimoNiskanen** [comment]
I will take a look at it, see if it is sufficiently straightforward to change to a bag table...

**Comment by RaimoNiskanen** [comment]
That was not entirely straightforward.

I have reverted to a bag table containing `#dns_rr{}` records, but do the lookups directly through ETS.
I suppose there should be no or minimal cache consistency issues, and I hope the performance is adequate.

Unfortunately I do not have good and thorough test suites, so it would be nice if others could review and test this for correctness and performance.

The next to try would be `{{Data, Class, Type} = Key, #dns_rr{}}` tuples to facilitate ETS lookups by `Key`.
After that `{{Data, Class, Type} = Key, [#dns_rr{}]}` tuples in a set table, but then simultaneous updates could delete each others' updates.

I wrote a long commit comment about these things.

**Comment by josevalim** [comment]
The commit looks great... but it also quite more complex! I wonder
if that's a reasonable price to pay for decentralization.

---

Regarding the commit message:

```
Try again to use a set table?
-----------------------------
Faster lookups would be achieved by storing
{{Domain, Class, Type} = Key, [#dns_rr{}]} tuples in a set table.

Updates of the cnt value could then be done by overwriting
the tuple with a new [#dns_rr{}] list.  Then parallel processes
might overwrite each others' changes, which might accedentally
delete a cache record written by a recent parallel lookup.
This might actually be acceptable since it only would cause
another cache miss.
```

What happens if we move out the `cnt` from the `#dns_rr{}` tuple
and keep it on the tuple too? Or is each `#dns_rr` record meant
to be expired/tracked individually?

I guess another approach is to have two tables. One in the format of
`{Key, Cnt}` (or maybe {Key, Tm, Cnt}) and another one keeping the
actual entries.

**Comment by RaimoNiskanen** [comment]
@josevalim 
> I wonder if that's a reasonable price to pay for decentralization.

If the performance gain is considerable, I think so.

Note that for lookup it is the new `match_rr/6` to compare with the old `filter_rr/2`, and it is a bit more complicated, but not very much (says the one that just wrote it and has the code fresh in the mind ;-).
Much of the other changes are me re-exploring Match Specifications to get fewer and more efficient passes through the table, so they have not much to do with decentralization.

> What happens if we move out the cnt from the #dns_rr{} tuple and keep it on the tuple too? Or is each #dns_rr record meant to be expired/tracked individually?

I left that out of my discussion in the commit comment.  That would enable using `ets:update_element/3`, and would remove the need to delete/insert tuples by the client code since the chief reason is to update `cnt` field, but it would cause subtle behaviour changes, maybe for the better:

Today they each `#dns_rr{}` record is expired individually.  The question is if it has to be so.

Records are cached by `res_cache_answer/1` that does a `lists:foreach/2` over all RRs, so all RRs from a reply are added "simultaneously", in a tight loop that adds them one at the time.  This tight loop could be rewritten to instead add a list of RRs as one entry with the same cache usage timestamp (`cnt` time).  Each RR has got its own `ttl` field so that suggests they should be expired individually.

However, how correct is the current behaviour?  If one RR should expire it is removed from the cache, but as long as there still are RRs that match the query they are returned as the complete answer even though a new query most probably would return also the expired RR with an updated TTL.  That does not sound correct to me.  Maybe it would be more correct to have both the `cnt` field and the `ttl` field (lowest of all RRs) in the outer tuple, and expire the whole RRs list entry.

This is a cache, maybe it should cache replies, not RRs?

> I guess another approach is to have two tables. One in the format of {Key, Cnt} (or maybe {Key, Tm, Cnt}) and another one keeping the actual entries.

Two tables often cause inconsistency problems...  Again, is it right of the current code to handle RRs individually?

**Comment by RaimoNiskanen** [comment]
@sneako, @josevalim: What do you think: to finish this PR; should we revert 76a2300 and b972eb7, and start over with new PR:s after 24.0, or should we in the name of performance accept this PR with my bag table fixes (and probably continue after 24.0 with a new PR for a set table and adjusted cache behaviour)?

**Comment by sneako** [comment]
@RaimoNiskanen I should be able to properly test out this PR early next week if that is soon enough. However, if this issue is the only thing blocking 24.0 at the moment, then I think we should probably revert those 2 commits and let everyone else start enjoying the new JIT while we continue working on this.

**Comment by RaimoNiskanen** [comment]
@sneako: we have a week or two left to code freeze for OTP-24

**Comment by sneako** [comment]
Ok, in that case I'll test this out ASAP :)

**Comment by RaimoNiskanen** [comment]
Early next week is just fine

**Comment by sneako** [comment]
Hey @RaimoNiskanen I started testing out the latest changes here and ran in to this: 
```
9> inet_tcp:getaddrs('kinesis.us-east-1.amazonaws.com').
{ok,[{3,91,171,130}]}
10> inet_tcp:getaddrs('kinesis.us-east-1.amazonaws.com').
{ok,[{3,91,171,130}]}
11> inet_tcp:getaddrs('kinesis.us-east-1.amazonaws.com').
** exception error: bad argument
     in function  ets:delete_object/2
        called as ets:delete_object(inet_cache,[])
        *** argument 2: not a tuple
     in call from inet_db:'-match_rr/6-lc$^1/1-1-'/2 (inet_db.erl, line 1698)
     in call from inet_db:match_rr/6 (inet_db.erl, line 1698)
     in call from inet_db:lookup_type/2 (inet_db.erl, line 690)
     in call from inet_db:hostent_by_domain/4 (inet_db.erl, line 669)
     in call from inet_db:getbyname/2 (inet_db.erl, line 612)
     in call from inet_res:getbyname_tm/3 (inet_res.erl, line 485)
     in call from inet:gethostbyname_tm/4 (inet.erl, line 1259)
```

Please let me know if you need any more information to help reproduce it

**Comment by RaimoNiskanen** [comment]
We need better test coverage...  The stacktrace was enough - I did not need to reproduce it.  Sorry about such a stupid bug!

**Comment by sneako** [comment]
No worries! Thanks for the quick fix, I will continue testing!

**Comment by RaimoNiskanen** [comment]
@sneako: We have code freeze for 24.0-rc3 today, so, in your opinion does it look promising enough to merge in now (in 1..2 hours) to further improve for 24.0, if necessary, or should I remove it from RC3 (and still validate, improve and probably include it in 24.0)?

**Comment by sneako** [comment]
@RaimoNiskanen I have had this running on a subset of our prod traffic for almost 1 hour now. Performance looks good and I have not encountered any further errors so far. I am a bit hesitant to rush this in however, since it took several days before we encountered #4631.

**Comment by RaimoNiskanen** [comment]
No stress.  Our release general lifted all branches (including this) out - it will not be in RC1, we will not rush it in.

Great to hear that the performance looks good/adequate!  Awaiting solid results...

**Comment by sneako** [comment]
I let this run on a subset of traffic for ~24hrs and it looked good, so now I am rolling it out to our entire cluster 👍  Will report back in another day or two

**Comment by sneako** [comment]
Things are still looking good! I'd say this PR is good to merge 😄

**Comment by RaimoNiskanen** [comment]
We will continue from here, then.
There are still many possibilities for optimization (and perhaps correction).

**Comment by RaimoNiskanen** [comment]
Thank you for your pull request!

**Pull request merged**